### ISSUE TITLE
getting error named "Segmentation fault" while evaluating the llm model

### ISSUE THREAD

**Issue by arunsingh2008** [mlflow/mlflow#10847]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Other

### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### MLflow version

(base) aion@ip-172-31-53-213:~$ pip show mlflow
Name: mlflow
Version: 2.9.2
Summary: MLflow: A Platform for ML Development and Productionization
Home-page: https://mlflow.org/
Author: Databricks
Author-email:
License: Apache License 2.0
Location: /home/aion/.local/lib/python3.10/site-packages
Requires: alembic, click, cloudpickle, databricks-cli, docker, entrypoints, Flask, gitpython, gunicorn, importlib-metadata, Jinja2, markdown, matplotlib, numpy, packaging, pandas, protobuf, pyarrow, pytz, pyyaml, querystring-parser, requests, scikit-learn, scipy, sqlalchemy, sqlparse
Required-by:

### System information

- os: linux
- Python version: 3.10.9

### Describe the problem

2024/01/18 13:46:37 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
2024/01/18 13:46:37 INFO mlflow.models.evaluation.default_evaluator: Computing model predictions.
2024/01/18 13:46:49 INFO mlflow.models.evaluation.default_evaluator: Testing metrics on first row...
Segmentation fault

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```

try:
    import transformers
    import mlflow
    import torch
    from huggingface_hub import snapshot_download
    import accelerate
    import pandas as pd
    import numpy as np
except:
    #!pip install transformers mlflow torch huggingface_hub pandas accelerate
    import transformers
    import mlflow
    import torch
    from huggingface_hub import snapshot_download
    import accelerate
    import pandas as pd
    import numpy as np



# model_loc = "/home/aion/arun" # change me
model_loc = "/home/aion/ptm/LLaMA-2-Chat/7B"
local_dir = "LLaMA-2-7B" # change me
experiment_name = "LLaMA-2-7B-instruct-evaluation" #change me




snapshot_location = model_loc
# snapshot_location = snapshot_download(repo_id="meta-llama/Llama-2-7b-hf", local_dir="llama-7b")

#Defining the Custom PyFunc
'''
Now, let's define our custom pyfunc. This will dictate how our model loads its dependencies and how 
it performs predictions. Notice how we've wrapped the intricacies of the model within this class.
'''

class MPT(mlflow.pyfunc.PythonModel):
    def load_context(self, context):
        """
        This method initializes the tokenizer and language model
        using the specified model snapshot directory.
        """
        # Initialize tokenizer and language model
        
        self.tokenizer = transformers.AutoTokenizer.from_pretrained(
            context.artifacts["snapshot"], padding_side="left"
        )

        config = transformers.AutoConfig.from_pretrained(
            context.artifacts["snapshot"], trust_remote_code=True
        )
        # If you are running this in a system that has a sufficiently powerful GPU with available VRAM,
        # uncomment the configuration setting below to leverage triton.
        # Note that triton dramatically improves the inference speed performance

        # config.attn_config["attn_impl"] = "triton"

        self.model = transformers.AutoModelForCausalLM.from_pretrained(
            context.artifacts["snapshot"],
            config=config,
            torch_dtype=torch.bfloat16,
            trust_remote_code=True,
        )

        # NB: If you do not have a CUDA-capable device or have torch installed with CUDA support
        # this setting will not function correctly. Setting device to 'cpu' is valid, but
        # the performance will be very slow.
        #self.model.to(device="cpu")
        # If running on a GPU-compatible environment, uncomment the following line:
        self.model.to(device="cuda")

        self.model.eval()

    def _build_prompt(self, instruction):
        """
        This method generates the prompt for the model.
        """
        INSTRUCTION_KEY = "### Instruction:"
        RESPONSE_KEY = "### Response:"
        INTRO_BLURB = (
            "Below is an instruction that describes a task. "
            "Write a response that appropriately completes the request."
        )

        return f"""{INTRO_BLURB}
        {INSTRUCTION_KEY}
        {instruction}
        {RESPONSE_KEY}
        """

    def predict(self, context, model_input, params=None):
        """
        This method generates prediction for the given input.
        """
        prompt = model_input["prompt"][0]

        # Retrieve or use default values for temperature and max_tokens
        temperature = params.get("temperature", 0.1) if params else 0.1
        max_tokens = params.get("max_tokens", 1000) if params else 1000

        # Build the prompt
        prompt = self._build_prompt(prompt)

        # Encode the input and generate prediction
        # NB: Sending the tokenized inputs to the GPU here explicitly will not work if your system does not have CUDA support.
        # If attempting to run this with GPU support, change 'cpu' to 'cuda' for maximum performance
        encoded_input = self.tokenizer.encode(prompt, return_tensors="pt").to(device="cuda")
        output = self.model.generate(
            encoded_input,
            do_sample=True,
            temperature=temperature,
            max_new_tokens=max_tokens,
        )

        # Decode the prediction to text
        generated_text = self.tokenizer.decode(output[0], skip_special_tokens=True)

        # Removing the prompt from the generated text
        prompt_length = len(self.tokenizer.encode(prompt, return_tensors="pt")[0])
        generated_response = self.tokenizer.decode(
            output[0][prompt_length:], skip_special_tokens=True
        )

        return {"candidates": [generated_response]}
# Building the Prompt


import pandas as pd
import numpy as np
import mlflow
from mlflow.models.signature import ModelSignature
from mlflow.types import DataType, Schema, ColSpec, ParamSchema, ParamSpec

# Define input and output schema
input_schema = Schema(
    [
        ColSpec(DataType.string, "prompt"),
    ]
)
output_schema = Schema([ColSpec(DataType.string, "candidates")])

parameters = ParamSchema(
    [
        ParamSpec("temperature", DataType.float, np.float32(0.1), None),
        ParamSpec("max_tokens", DataType.integer, np.int32(1000), None),
    ]
)

signature = ModelSignature(inputs=input_schema, outputs=output_schema, params=parameters)


# Define input example
input_example = pd.DataFrame({"prompt": ["What is machine learning?"]})

#Set the experiment that we're going to be logging our custom model to

'''
If the experiment doesn't already exist, MLflow will create a new experiment 
with this name and will alert you that it has created a new experiment.'''

mlflow.set_experiment(experiment_name= experiment_name)

# Get the current base version of torch that is installed, without specific version modifiers
torch_version = torch.__version__.split("+")[0]



#breakpoint()
eval_data = pd.DataFrame(
    {
        "prompt": [
            "What is MLflow?",
            "What is Spark?",
        ],
        "ground_truth": [
            "MLflow is an open-source platform for managing the end-to-end machine learning (ML) "
            "lifecycle. It was developed by Databricks, a company that specializes in big data and "
            "machine learning solutions. MLflow is designed to address the challenges that data "
            "scientists and machine learning engineers face when developing, training, and deploying "
            "machine learning models.",
            "Apache Spark is an open-source, distributed computing system designed for big data "
            "processing and analytics. It was developed in response to limitations of the Hadoop "
            "MapReduce computing model, offering improvements in speed and ease of use. Spark "
            "provides libraries for various tasks such as data ingestion, processing, and analysis "
            "through its components like Spark SQL for structured data, Spark Streaming for "
            "real-time data processing, and MLlib for machine learning tasks",
        ],
    }
)
eval_data = eval_data.head(1)

with mlflow.start_run():
    logged_model_info = mlflow.pyfunc.log_model(
        experiment_name,
        python_model=MPT(),
        # NOTE: the artifacts dictionary mapping is critical! This dict is used by the load_context() method in our MPT() class.
        artifacts={"snapshot": snapshot_location},
        pip_requirements=[
            f"torch=={torch_version}",
            f"transformers=={transformers.__version__}",
            f"accelerate=={accelerate.__version__}",
            "einops",
            "sentencepiece",
        ],
        input_example=input_example,
        signature=signature,
    )
    # Use predefined question-answering metrics to evaluate our model.
    breakpoint()
    results = mlflow.evaluate(
        logged_model_info.model_uri,
        eval_data,
        targets="ground_truth",
        model_type="question-answering",
    )
    print(f"See aggregated evaluation results below: \n{results}")
    # print(f"See aggregated evaluation results below: \n{results.metrics}")

    # Evaluation result for each data record is available in `results.tables`.
    eval_table = results.tables["eval_results_table"]
    print(f"See evaluation table below: \n{eval_table}")


```

### Stack trace

```
REPLACE_ME
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by WeichenXu123** [comment]
Thanks for reporting! I will investigate this.

**Comment by MariaFjodorowa** [comment]
The issue may be related to the rouge metrics. In my case, jupyter kernel dies silently even when running rouge only on a 5 rows long eval_df without any LLM in memory:


```
mlflow.evaluate(
    data=eval_df.head(),
    targets="ground_truth",
    predictions="predictions",
    extra_metrics=[
    mlflow.metrics.rouge1(),
    ],
)
```


but other metrics e.g. mlflow.metrics.ari_grade_level() or a custom unbabel-comet metric work normally.

**Comment by arunsingh2008** [comment]
@MariaFjodorowa Thanks...so do I need to remove Rouge metrics
Please give me the code for the same

**Comment by WeichenXu123** [comment]
Thanks ! I think this should not be mlflow side issue. MLflow is a pure python library and it is unlikely becomes root cause of a "Segmentation fault".

I think  you can remove mlflow related stuff from your repro code and the issue should still persist, if so you can report this bug to "transformers" project. :)

**Comment by MariaFjodorowa** [comment]
@arunsingh2008 you can do so by not specifying model_type in mlflow.evaluate and replacing mlflow.metrics.rouge1() with metrics you need in extra_metrics in my example above

**Comment by arunsingh2008** [comment]
(base) aion@ip-172-31-50-243:~$ python llm/llm_mlflow_logging.py
Downloading artifacts: 100%|█████████████████████████████████████████████████████████████████████████████| 9/9 [01:00<00:00,  6.76s/it]
2024/01/23 11:35:19 WARNING mlflow.pyfunc: Detected one or more mismatches between the model's dependencies and the current Python environment:
 - sentencepiece (current: uninstalled, required: sentencepiece)
To fix the mismatches, call `mlflow.pyfunc.get_model_dependencies(model_uri)` to fetch the model's environment and install dependencies using the resulting environment file.
Loading checkpoint shards:   0%|                                                                                 | 0/2 [00:00<?, ?it/s]/home/aion/.local/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████| 2/2 [00:01<00:00,  1.76it/s]
/home/aion/.local/lib/python3.10/site-packages/mlflow/data/digest_utils.py:26: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_columns = trimmed_df.columns[(df.applymap(type) == str).all(0)]
/home/aion/.local/lib/python3.10/site-packages/mlflow/models/evaluation/base.py:414: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
2024/01/23 11:35:23 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
2024/01/23 11:35:23 INFO mlflow.models.evaluation.default_evaluator: Computing model predictions.
2024/01/23 11:35:28 INFO mlflow.models.evaluation.default_evaluator: Testing metrics on first row...
Segmentation fault
(base) aion@ip-172-31-50-243:~$



For other metrices also I am getting same error

**Comment by arunsingh2008** [comment]
@MariaFjodorowa Please support on above.. I am getting same issue for other metrics as well
